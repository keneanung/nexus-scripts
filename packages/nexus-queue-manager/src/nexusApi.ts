// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

export const sendCommand = (command: string) => {
  nexusclient.send_commands(command, false);
};
