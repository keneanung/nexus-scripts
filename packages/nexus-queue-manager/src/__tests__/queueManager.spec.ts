import { QueueManager } from '..';
import { sendCommand } from '../nexusApi';

jest.mock('../nexusApi');
const sendCommandMock = jest.mocked(sendCommand);

// Reset handlers so that each test could alter them
// without affecting other, unrelated tests.
afterEach(() => {
  sendCommandMock.mockReset();
});

test.each([
  ['e'],
  ['b'],
  ['c'],
  ['s'],
  ['p'],
  ['w'],
  ['u'],
  ['t'],
  ['!e'],
  ['!b'],
  ['!c'],
  ['!s'],
  ['!p'],
  ['!w'],
  ['!u'],
  ['!t'],
  ['equilibrium'],
  ['balance'],
  ['class'],
  ['paralysis'],
  ['unbound'],
  ['stun'],
  ['free'],
  ['freestand'],
  ['full'],
  ['frl'],
])('Should translate property %p correctly', (queue: string) => {
  const sut = new QueueManager();

  sut.track('foo', queue);

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have an empty queue after clear("all")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('all');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have remove items in class queue after clear("class")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('class');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have remove items in class queue after clear("c")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('c');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should put prepended items first in queue', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackFirst('bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should put items at given position in queue with trackAt', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackAt(3, 'bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should replace items at given position in queue with trackReplace', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackReplace(2, 'bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove items at given position in queue with trackRemove', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackRemove(2);

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('foo', 'full');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue with different casing', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('FOO', 'full');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue (not first item)', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('foo', 'free');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Workaround for in-game bug #17807', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('stand', 'full');
  sut.track('sit', 'eu');

  sut.run('sit', 'eu');
  sut.run('stand', 'eu');
  sut.run('sit', 'eu');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should queue command to in-game queue when added locally and queue not full', () => {
  const sut = new QueueManager();

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });

  expect(sendCommandMock).toMatchSnapshot();
});

test('Should not attempt to queue command to in-game queue when added locally and queue is full', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });

  expect(sendCommandMock).toHaveBeenCalledTimes(0);
});

test('Should recognize own queue item on confirmation in game and mark it accordingly', () => {
  const sut = new QueueManager();

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });
  sut.track('stand', 'eb!p!t!w');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should not mark same item multiple times as local', () => {
  const sut = new QueueManager();

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });
  sut.track('stand', 'eb!p!t!w');
  sut.track('stand', 'eb!p!t!w');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should queue command to in-game queue when added locally and queue is full, but queue ran', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });
  sut.run('sit', 'full');

  expect(sendCommandMock).toMatchSnapshot();
});

test('Should not queue commands that are already being queued', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });
  sut.track('sit', 'full');
  sut.run('sit', 'full');

  expect(sendCommandMock).toMatchSnapshot();
});

test('Should re-attempt to queue things if first attempt was blocked due to queue being full', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');
  sut.track('sit', 'full');

  sut.do('stand', {
    haveBalance: true,
    haveEq: true,
    haveParalysis: false,
    beStunned: false,
    beBound: false,
  });
  sut.track('sit', 'full');
  sut.blocked();
  sut.run('sit', 'full');

  expect(sendCommandMock).toMatchSnapshot();
});

test('Should re-queue things after being queued with dor and run', () => {
  const sut = new QueueManager();

  sut.do(
    'stand',
    {
      haveBalance: true,
      haveEq: true,
      haveParalysis: false,
      beStunned: false,
      beBound: false,
    },
    true,
  );
  sut.track('stand', 'eb!p!t!w');
  sut.run('stand', 'eb!p!t!w');

  expect(sendCommandMock).toMatchSnapshot();
});
