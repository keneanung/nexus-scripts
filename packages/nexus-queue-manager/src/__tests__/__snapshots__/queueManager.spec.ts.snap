// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have an empty queue after clear("all") 1`] = `[]`;

exports[`Should have remove items in class queue after clear("c") 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should have remove items in class queue after clear("class") 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should not mark same item multiple times as local 1`] = `
[
  {
    "command": "stand",
    "locallyControlled": true,
    "properties": {
      "beBound": false,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "stand",
    "locallyControlled": false,
    "properties": {
      "beBound": false,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should not queue commands that are already being queued 1`] = `
[MockFunction] {
  "calls": [
    [
      "queue add eb!p!w!t stand",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should put items at given position in queue with trackAt 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "bar",
    "locallyControlled": false,
    "properties": {
      "haveClassBalance": true,
    },
    "repeat": false,
  },
  {
    "command": "bamboozle",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should put prepended items first in queue 1`] = `
[
  {
    "command": "bamboozle",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "bar",
    "locallyControlled": false,
    "properties": {
      "haveClassBalance": true,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should queue command to in-game queue when added locally and queue is full, but queue ran 1`] = `
[MockFunction] {
  "calls": [
    [
      "queue add eb!p!w!t stand",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should queue command to in-game queue when added locally and queue not full 1`] = `
[MockFunction] {
  "calls": [
    [
      "queue add eb!p!w!t stand",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should re-attempt to queue things if first attempt was blocked due to queue being full 1`] = `
[MockFunction] {
  "calls": [
    [
      "queue add eb!p!w!t stand",
    ],
    [
      "queue add eb!p!w!t stand",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should re-queue things after being queued with dor and run 1`] = `
[MockFunction] {
  "calls": [
    [
      "queue add eb!p!w!t stand",
    ],
    [
      "queue add eb!p!w!t stand",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should recognize own queue item on confirmation in game and mark it accordingly 1`] = `
[
  {
    "command": "stand",
    "locallyControlled": true,
    "properties": {
      "beBound": false,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should remove items at given position in queue with trackRemove 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should remove run command from queue (not first item) 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "bar",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "boom",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should remove run command from queue 1`] = `
[
  {
    "command": "bar",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "boom",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should remove run command from queue with different casing 1`] = `
[
  {
    "command": "bar",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "boom",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should replace items at given position in queue with trackReplace 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "bamboozle",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
  {
    "command": "baz",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!b" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveBalance": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!c" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveClassBalance": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!e" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveEq": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!p" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!s" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!t" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!u" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "!w" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beBound": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "b" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "balance" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "c" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveClassBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "class" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveClassBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "e" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveEq": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "equilibrium" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveEq": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "free" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "freestand" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "frl" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {},
    "repeat": false,
  },
]
`;

exports[`Should translate property "full" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "p" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveParalysis": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "paralysis" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveParalysis": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "s" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "haveShipBalance": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "stun" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "t" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStunned": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "u" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beStanding": true,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "unbound" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beBound": false,
    },
    "repeat": false,
  },
]
`;

exports[`Should translate property "w" correctly 1`] = `
[
  {
    "command": "foo",
    "locallyControlled": false,
    "properties": {
      "beBound": true,
    },
    "repeat": false,
  },
]
`;

exports[`Workaround for in-game bug #17807 1`] = `[]`;
